src/Poseidon/EntitiesList.hs

```diff
@@ -39,63 +40,98 @@ instance Show SignedEntity where
 
 type SignedEntitiesList = [SignedEntity]
 
-isInclude :: SignedEntity -> Bool
-isInclude (Include _) = True
-isInclude _ = False
-
-isExclude :: SignedEntity -> Bool
-isExclude (Exclude _) = True
-isExclude _ = False
-
-fromSignedEntity :: SignedEntity -> PoseidonEntity
-fromSignedEntity (Include a) = a
-fromSignedEntity (Exclude a) = a
-
-entityIncludes :: SignedEntitiesList -> EntitiesList
-entityIncludes xs = map fromSignedEntity $ filter isInclude xs
-
-entityExcludes :: SignedEntitiesList -> EntitiesList
-entityExcludes xs = map fromSignedEntity $ filter isExclude xs
-
--- | A parser to parse entities
-parseSignedEntitiesList :: P.Parser SignedEntitiesList
-parseSignedEntitiesList = do 
-    eL <- P.sepBy parsePoseidonEntity (P.char ',' <* P.spaces)
-    P.optional (P.many (P.char ' ') >> comment)
-    return eL
-
-parseSignedEntitiesListFromFile :: P.Parser SignedEntitiesList
-parseSignedEntitiesListFromFile = do
-    concat <$> P.sepBy (P.try emptyLine <|> comment <|> parseSignedEntitiesList) P.newline
+-- A class to generalise signed and unsigned Entity Lists. Both have the feature that they can be used to filter individuals.
+class EntitySpec a where
+    indInfoConformsToEntitySpec :: [a] -> IndividualInfo -> Bool
+    underlyingEntity :: a -> PoseidonEntity
+    entitySpecParser :: P.Parser a
+
+instance EntitySpec SignedEntity where
+    indInfoConformsToEntitySpec signedEntities (IndividualInfo indName groupNames pacName) = go signedEntities False
+      where
+        go [] r = r
+        go (Include (Ind   n):rest) r = if n ==     indName    then go rest True  else go rest r
+        go (Include (Group n):rest) r = if n `elem` groupNames then go rest True  else go rest r
+        go (Include (Pac   n):rest) r = if n ==     pacName    then go rest True  else go rest r
+        go (Exclude (Ind   n):rest) r = if n ==     indName    then go rest False else go rest r
+        go (Exclude (Group n):rest) r = if n `elem` groupNames then go rest False else go rest r
+        go (Exclude (Pac   n):rest) r = if n ==     pacName    then go rest False else go rest r
```

Comment:
I'm not sure about this recursive solution. Wouldn't a fold with a helper function for the comparison be more readable? Or do I miss something here?

Reply:
Yeah, I guess a `fold` could also work, but I find recursions to be very natural in Haskell now, and I learned thinking in terms of them more easily than in `fold`s. I basically wouldn't be able to cast this into a fold easily, I believe. Happy for suggestions, though.

Reply:
I implemented my solution in #165.